File: micros/maxinoutdegree.cc
/**
 * Compute the indegree and outdegree of all nodes in a graph and
 * print the maximums.
 */

#include <stdio.h>
#include <string.h>
#include <iostream>
#include "pmgd.h"
#include "util.h"
#include "time.h"

using namespace std;
using namespace PMGD;

/**
 * Program name.
 */
static const char prog_name[] = "maxinoutdegree";

/**
 * Print information about how to invoke this program
 * @param o the put-to stream
 */
void print_usage(ostream &o)
{
    o << "Usage: " << prog_name << " [OPTION]... GRAPH\n";
    o << "\n";
    o << "Find the nodes in GRAPH with the maximum indegree and outdegree.\n";
    o << "\n";
    o << "  -h  print this help and exit\n";
    o << "  -t  print elapsed time and usage\n";
}

/**
 * The application's main function
 * @param argc the number of command line arguments
 * @param argv the array of command line arguments
 * @return status of execution (0 okay, 1 bad arguments, 2 processing error)
 */
int main(int argc, char *argv[])
{
    bool timing = false;
    struct timeval t1, t2;
    struct rusage u1, u2;

    char *graph_name;

    long long max_indegree;
    long long max_outdegree;

    // Parse the command line arguments
    int argi = 1;
    while (argi < argc && argv[argi][0] == '-') {
        if (strcmp(argv[argi], "-h") == 0) {
            print_usage(cout);
            return 0;
        }
        else if (strcmp(argv[argi], "-t") == 0) {
            timing = true;
        }
        else {
            fprintf(stderr, "%s: %s: Unrecognized option\n", prog_name, argv[argi]);
            return 1;
        }
        argi += 1;
    }
    if (argi + 1 > argc) {
        cerr << prog_name << ": No input graph\n";
        print_usage(cerr);
        return 1;
    }
    graph_name = argv[argi];

    try {
        // Open the graph.
        Graph db(graph_name, Graph::ReadOnly);

        // Start a transaction.
        Transaction tx(db);

        // Time and usage measurements.
        (void)getrusage(RUSAGE_SELF, &u1);
        (void)gettimeofday(&t1, NULL);

        // Compute the indegree and outdegree of each node, and
        // compute the maximums.
        max_indegree = -1;
        max_outdegree = -1;
        for (NodeIterator n = db.get_nodes(); n; n.next()) {
            long long indegree = 0;
            for (EdgeIterator e = n->get_edges(Incoming); e; e.next())
                ++indegree;
            if (indegree > max_indegree) max_indegree = indegree;

            long long outdegree = 0;
            for (EdgeIterator e = n->get_edges(Outgoing); e; e.next())
                ++outdegree;
            if (outdegree > max_outdegree) max_outdegree = outdegree;
        }

        // Time and measurements.
        (void)gettimeofday(&t2, NULL);
        (void)getrusage(RUSAGE_SELF, &u2);
    }
    catch (Exception e) {
        print_exception(e, stderr);
        return 2;
    }

    cout << max_indegree << "\n";
    cout << max_outdegree << "\n";

    if (timing) print_delta(cout, t1, t2, u1, u2);

    return 0;
}

File: benchmarks/microbenchmarks/MaxInOutDegree.java
/**
 * Compute the indegree and outdegree of all nodes in a graph and
 * print the maximums.
 */
import pmgd.*;
import java.lang.String;
import java.lang.management.*;

public class MaxInOutDegree {
    /**
     * Program name.
     */
    static String progName = "MaxInOutDegree";

    /**
     * The application's main function
     * @param argc the number of command line arguments
     * @param argv the array of command line arguments
     * @return status of execution (0 okay, 1 bad arguments, 2 processing error)
     */
    public static void  main(String[] argv)
    {
        boolean timing = false;
        int argc = argv.length;

        String graph_name;

        long max_indegree;
        long max_outdegree;

        // Parse the command line arguments
        int argi = 0;
        while (argi < argc && argv[argi].charAt(0) == '-') {
            if (argv[argi].compareTo("-h") == 0) {
                printUsage();
                return;
            }
            else if (argv[argi].compareTo("-t") == 0) {
                timing = true;
            }
            else {
                System.out.println(progName + ": " + argv[argi] + ": Unrecognized option");
                return;
            }
            argi += 1;
        }
        if (argi + 1 > argc) {
            System.out.println(progName + ": No input graph");
            printUsage();
            return;
        }
        graph_name = argv[argi];

        ThreadMXBean bean = ManagementFactory.getThreadMXBean();
        long userTime, cpuTime;

        try {
            // Open the graph.
            Graph db = new Graph(graph_name, Graph.OpenOptions.ReadOnly);

            // Start a transaction.
            Transaction tx = new Transaction(db, false, true);

            // Time measurement.
            long startTime = bean.getCurrentThreadUserTime();
            long startTotal = bean.getCurrentThreadCpuTime();

            // Compute the indegree and outdegree of each node, and
            // compute the maximums.
            max_indegree = -1;
            max_outdegree = -1;
            for (NodeIterator n = db.get_nodes(); !n.done(); n.next()) {
                long indegree = 0;
                for (EdgeIterator e = n.get_current().get_edges(Node.Direction.Incoming); !e.done(); e.next())
                    ++indegree;
                if (indegree > max_indegree) max_indegree = indegree;

                long outdegree = 0;
                for (EdgeIterator e = n.get_current().get_edges(Node.Direction.Outgoing); !e.done(); e.next())
                    ++outdegree;
                if (outdegree > max_outdegree) max_outdegree = outdegree;
            }

            // Time measurement.
            long endTime = bean.getCurrentThreadUserTime();
            long endTotal = bean.getCurrentThreadCpuTime();
            userTime = (endTime - startTime)/1000;
            cpuTime = (endTotal - startTotal)/1000;
        }
        catch (pmgd.Exception e) {
            e.print();
            return;
        }

        System.out.println(max_indegree);
        System.out.println(max_outdegree);

        if (timing) {
            long sysTime = cpuTime - userTime;
            float upercent = 100 * (userTime / (float)cpuTime);
            float spercent = 100 * (sysTime / (float)cpuTime);
            System.out.println("Elapsed time is " + cpuTime + " microseconds");
            System.out.println("User time is " + userTime + " microseconds (" + upercent + "%)");
            System.out.println("System time is " + sysTime + " microseconds (" + spercent + "%)");
        }

        return;
    }

    /**
     * Print information about how to invoke this program
     * @param o the put-to stream
     */
    static void printUsage()
    {
        System.out.println("Usage: " + progName + " [OPTION]... GRAPH");
        System.out.println("Find the nodes in GRAPH with the maximum indegree and outdegree.");
        System.out.println("  -h  print this help and exit");
        System.out.println("  -t  time the execution");
    }
}
